### PR TITLE
Enable RT-DETR CoreML export and protect decoder ops from FP16 drift

### PR DESCRIPTION
**Pull request by artest08**
RT-DETR-L now exports to CoreML mlprogram via three traceability fixes (AIFI position embedding dtype, MSDeformAttn rewritten with rank-5 tensors for MIL's rank limit, integer division via `torch.div(rounding_mode="floor")`), and the exporter keeps `linear` and `gather*` ops in FP32 so the decoder's score-head matmuls and deformable-sampling indices retain precision while the rest of the graph stays FP16. Validated at 52.8 mAP on COCO val with batch=1 with rtdetr-l.

```bash
yolo export model=./rtdetr-l.pt format=coreml opset=17 half=True batch=1
yolo val model=./rtdetr-l.mlpackage data=coco.yaml half=True batch=1
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🚀 This PR improves **RT-DETR CoreML export support** in Ultralytics by updating model code, export settings, and tests to make CoreML conversion more reliable and accurate.

### 📊 Key Changes
- 🧪 Added a new test for exporting **RT-DETR** models to **CoreML**, including a macOS inference check after export.
- 🔧 Updated RT-DETR postprocessing logic to avoid integer operations that CoreML does not handle well during conversion.
- 🎯 Adjusted transformer position embedding calculations to stay in **float32** where needed, reducing accuracy issues during CoreML export.
- 📦 Refactored deformable attention code to keep tensor ranks lower, which matches **CoreML MIL** converter limits and avoids export failures.
- 🧠 Simplified attention tensor shaping by folding multiple dimensions together while preserving the same model behavior.
- ⚙️ Updated CoreML export settings so **specific RT-DETR operations stay in float32**, preventing precision drift in exported models.
- 📝 Improved code comments and docstrings to clarify CoreML-related constraints and broadcasting behavior.

### 🎯 Purpose & Impact
- ✅ Makes **RT-DETR CoreML export work more reliably**, especially for Apple deployment workflows.
- 🍎 Improves compatibility with **CoreML tools and Apple platforms**, reducing conversion errors and unsupported-operation issues.
- 🎯 Helps preserve **prediction quality** by avoiding precision loss in sensitive RT-DETR layers.
- 🛠️ Strengthens test coverage, so future changes are less likely to break RT-DETR CoreML export silently.
- 👥 Benefits developers deploying Ultralytics models to iPhone, iPad, and macOS apps, while keeping standard YOLO export behavior unchanged.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @artest08, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an engineer will assist you shortly. Please review the checklist below to help ensure a smooth integration. 😊

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/export/coreml.py [hunk 1]
```diff
@@ -198,17 +198,23 @@ def torch2coreml(
...
         inputs=inputs,
         classifier_config=ct.ClassifierConfig(classifier_names) if classifier_names else None,
         convert_to="neuralnetwork" if mlmodel else "mlprogram",
         skip_model_load=True,
     )
+    if not mlmodel:
+        # Keep linear and gather variants in fp32 to preserve RT-DETR decoder accuracy: linear matmuls
+        # in the score head shift class logits, and fp16 gather/gather_nd in deformable-attention
+        # sampling and top-k indexing perturbs which features feed each query.
+        fp32_ops = {"linear", "gather", "gather_nd", "gather_along_axis"}
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/export/coreml.py]: 💡 **MEDIUM**: This selector still leaves the decoder attention `softmax` in FP16, so one of the main drift-sensitive steps is not actually protected even though the surrounding comment says the attention softmax path is part of the fp32 carve-out. That can still move RT-DETR query rankings on CoreML exports.

**Suggested change:**
```suggestion
        fp32_ops = {"linear", "softmax", "gather", "gather_nd", "gather_along_axis"}
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The CoreML export and RT-DETR decoder changes look well scoped overall, but the mixed-precision guard currently misses `softmax`, so the PR does not fully protect the decoder path it calls out as numerically sensitive.

💬 Posted 1 inline comment

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-05-16 15:30:34 UTC with `@ultralytics/run-ci` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/25965733701): `ci.yml` run 22007

**Comment by Laughing-q** [comment]
@artest08 is this PR ready to go?

**Comment by artest08** [comment]
@Laughing-q Actually I have refined one more time, I have restrained linear pinning only for rtdetrdecoder in this refinement, because otherwise it might lead to undesired fp32 pinning for also yolo models.

And I have tested this only with specific coreml package and on my mac. 

With the commands below, I am obtaining a validation performance of 52.7 with rtdetr-l

yolo export model=./rtdetr-l.pt format=coreml opset=17 half=True batch=1
yolo val model=./rtdetr-l.mlpackage data=coco.yaml half=True batch=1

However, a second eye to the PR can also be good. I just feel that I might be missing some points.

**Review by Laughing-q — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
Merged! 🎉

Huge thanks to @artest08, with contributions from @onuralpszr and @Laughing-q, for strengthening RT-DETR CoreML export support. This is a meaningful improvement for Apple deployment workflows, from more reliable conversion and better float32 handling to stronger test coverage that helps protect future stability.

As Leonardo da Vinci said, “Simplicity is the ultimate sophistication.” This PR reflects that well by simplifying critical tensor and postprocessing paths while improving accuracy, compatibility, and reliability where it matters most.

Really appreciate the thoughtful work here and the impact it brings to the Ultralytics community.
